### DISCUSSION TITLE
How to get Keycloak's roles?

### DISCUSSION THREAD

**Discussion by LoiKos** [nextauthjs/next-auth#8584] [Help] [answered]
I wanted to come back on this issue : https://github.com/nextauthjs/next-auth/issues/4811

I check both `account` & `profile` as recommended and i can find a lot of informations like `access_token`, `refresh_token` but roles are missing.

The only way i found to get the roles are to parsed the token myself : 

```ts
import jwt from 'jsonwebtoken'
//...
 callbacks: {
    //...
    jwt: async (params) => {
      let roles = undefined as
        | { [key: string]: { roles: Array<any> } }
        | undefined

      if (account?.access_token) {
        let decodedToken = jwt.decode(account?.access_token)
        if (decodedToken && typeof decodedToken !== 'string') {
          roles = decodedToken?.resource_access
        }
      }
      //...
    }
 }
```

I'm not really comfortable with this approach and i'm wondering why it's missing from `profile`.
`profile` and `decodedToken` all fields are similar except for `aud`. `profile` is missing the following fields: `realm_access` and `resource_access` and `scope`.

Is this happening because other providers doesn't define these fields ?

**Comment by kwame-mintah** [comment, marked as answer]
@LoiKos, what version of Keycloak are you using? There is a way to include the roles as part of the access token.

**Reply by LoiKos** [reply]
I'm using keycloak version 19.x (not the latest). I'm aware of that and in fact roles are correctly defined in the access_token and i can retrieve them by decoding my self the token. But i don't understand why their not part of profile when profile seems to correspond to a "parsed" access_token.

**Reply by LoiKos** [reply]
That said i use nex-auth v5 which is experimental and it might just be a bug.

**Reply by kwame-mintah** [reply]
I would say it's expected behaviour, as roles isn't part of the standard claims:
https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

**Reply by LoiKos** [reply]
Probably ...Imo it's a bit sad `profile` doesn't send back the full parsed token.

**Reply by kwame-mintah** [reply]
The Keycloak scope for profile mapping matches the OpenID specs linked above. However, this doesn't stop you from amending the profile scopes to include roles within your instance.

1. Navigate to 'Client scopes' then select 'profile'
2. On the 'Mappers' tab click the 'Add mapper' -> 'From predefined mappers'
3. In the new window, find the 'realm roles'.

Hope this helps you.

**Reply by kwame-mintah** [reply]
@LoiKos - I have just tested this on my local instance and following the instructions above, you can get the roles to be returned when the scope is `profile`.

**Reply by LoiKos** [reply]
@kwame-mintah in fact it work correctly. My problem was not related to next-auth finally but with keycloak itself. I have a "." in my token claim name that cause it to not appearing in the token :/

**Reply by martinzokov** [reply]
Can you elaborate a bit on how you solved this please? I used your snippet like so:
```typescript
callbacks: {
    async jwt({token, user, account, profile }) {
      
      //console.log("account "+ JSON.stringify(account))
      let roles = undefined as
        | { [key: string]: { roles: Array<any> } }
        | undefined

      if (account?.access_token) {
        let decodedToken = jwt.decode(account?.access_token)
        if (decodedToken && typeof decodedToken !== 'string') {
          roles = decodedToken?.realm_access
        }
      }
      token = {...token, roles: roles?.roles}
      console.log("[jwt callback] token " + JSON.stringify(token))
      return token;
    },
    session: async ({ session, token, user }) => {
      if (session?.user) {
        session.user.id = token.sub;
        
      }
      console.log("[session callback] token " + JSON.stringify(token))

      return session;
    },
  },
```

However, in my session callback it seems to be passing only the original token without the roles.
Logs:
```
[jwt callback] token {"name":"demouser","sub":"ec710553-6f5e-4a2f-ba8e-c30159175c1d","roles":["doctor"]}
...
[session callback] token {"name":"demouser","sub":"ec710553-6f5e-4a2f-ba8e-c30159175c1d","iat":1715281935,"exp":1717873935,"jti":"05f3536f-2b4e-412a-bed8-5eece7dafc43"}
```

**Reply by LoiKos** [reply]
I check your snippet and it seems that you change `roles = decodedToken?.resource_access` to `roles = decodedToken?.realm_access`. Did you try with `resource_access` ?

**Reply by martinzokov** [reply]
Yes, I think my roles just get sent in the realm_access property. I actually found a solution to my issue. I copy/pasted the config and oidc.ts from this repo: [https://github.com/dkrasnovdev/nextjs-app-router-keycloak-example
](https://github.com/dkrasnovdev/nextjs-app-router-keycloak-example
)
Now it works for me!
